### PR TITLE
Compute output shapes for some kernels

### LINKED ISSUE #2368: Compute output shapes for some kernels
@tensorflow/micro 

Update the output tensor shape during prepare phase when the computed shape does not match the shape in the flatbuffer.

Kernels:
- BATCH_TO_SPACE_ND
- SPACE_TO_BATCH_ND
- CONV
- RESHAPE
- EXPAND_DIMS
- DEPTHWISE_CONV

Update CMSIS_NN and ARC_MLI optimized kernels.
Add additional tests from TfLite.
Update existing tests.

### PR DESCRIPTION
**Pull request by ddavis-2015**
@tensorflow/micro

Update the output tensor shape during prepare phase when the computed shape does not match the shape in the flatbuffer.

Kernels:
- BATCH_TO_SPACE_ND
- SPACE_TO_BATCH_ND
- CONV
- RESHAPE
- EXPAND_DIMS
- DEPTHWISE_CONV

Update CMSIS_NN and ARC_MLI optimized kernels.
Add additional tests from TfLite for BATCH_TO_SPACE_ND and SPACE_TO_BATCH_ND.
Update existing tests.
Add tests for Keras model using convolution with dilation > 1.

Update memory_arena_threshold_test to increase total, tail, and persistent allocation sizes:
- Add 20 bytes for CONV output shape
- Add 15 bytes for arena allocation alignment
- x2 convolution layers

Update micro_speech_test arena size as per description in C++ code.

See #2319 for additional details.

Resolves [b/317362237](https://issuetracker.google.com/317362237)

bug=fixes #2368 #1646 #1629 #1231 #2338 #2319

### CODE REVIEW

File: tensorflow/lite/micro/kernels/expand_dims_test.cc [hunk 1]
```diff
@@ -187,25 +184,41 @@ TF_LITE_MICRO_TEST(ExpandDimsNegativeAxisTest1) {
...
+  // tensor dimensions are computed correctly.
+  int golden_dims[] = {4, 1, 3, 2, 1};
   int output_dims[] = {4, 1, 3, 1, 2};
 
+  tflite::testing::TestExpandDims<float>(input_dims, input_data, axis_dims,
+                                         axis_data, golden_dims, output_dims,
+                                         golden_data, output_data);
+}
+
+TF_LITE_MICRO_TEST(ExpandDimsAxisOutOfRangeShallFailTest1) {
```
Comment (rascani) [inline tensorflow/lite/micro/kernels/expand_dims_test.cc:212]: Nit: rather than having these tests listed as 1 & 2, can we be more specific with the name about what failure condition it is triggering? Seems like 1 is too high of an axis value, while 2 is too low. Maybe ExpandDimsAxisOutOfRangeHighShallFailTest & ExpandDimsAxisOutOfRangeLowShallFailTest.

**Review by rascani — COMMENTED** [review]
This LGTM, just one tiny nit. We should probably file follow-up issues for supporting this with CEVA & Xtensa optimized kernels.

**Comment by ddavis-2015** [comment]
> This LGTM, just one tiny nit. We should probably file follow-up issues for supporting this with CEVA & Xtensa optimized kernels.

The CEVA and Xtensa code calls the reference Prepare method.  So no additional support work is required.

File: tensorflow/lite/micro/kernels/expand_dims_test.cc (hunk 1 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/kernels/expand_dims_test.cc]: done.

**Review by rascani — APPROVED** [review]
